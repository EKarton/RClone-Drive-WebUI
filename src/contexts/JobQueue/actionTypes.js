export const ActionTypes = Object.freeze({
  SET_JOBS: 'SET_JOBS',
  ADD_JOB: 'ADD_JOB',
  ADD_JOBS: 'ADD_JOBS',
  REMOVE_JOB: 'REMOVE_JOB',
  ADD_STATUS_COUNT: 'ADD_STATUS_COUNT',
  UPDATE_STATUS: 'UPDATE_STATUS',
});
