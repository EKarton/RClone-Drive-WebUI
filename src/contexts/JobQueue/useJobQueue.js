import RCloneJobTracker from 'services/RCloneJobTracker';

export const rCloneJobTracker = new RCloneJobTracker();
