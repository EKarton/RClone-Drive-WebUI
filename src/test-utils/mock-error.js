/**
 * Replaces the error object's stack trace with mock data
 * @param {Error} error the Error object
 * @returns {Error} the same error, but with the stack trace mocked
 */
export function mockErrorStackTrace(error) {
  error.stack = 'Mock content for stack trace';
  return error;
}
