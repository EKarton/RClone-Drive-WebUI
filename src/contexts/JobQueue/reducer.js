import { JobStatus } from 'utils/constants';
import { ActionTypes } from './actionTypes';

export function reducer(state, action) {
  switch (action.type) {
    case ActionTypes.SET_JOBS: {
      return { ...state, jobs: action.payload };
    }
    case ActionTypes.ADD_JOB: {
      const newJobs = [action.payload, ...state.jobs];

      return { ...state, jobs: newJobs };
    }
    case ActionTypes.ADD_JOBS: {
      const newJobs = [...action.payload, ...state.jobs];

      return { ...state, jobs: newJobs };
    }
    case ActionTypes.REMOVE_JOB: {
      const newJobs = state.jobs.filter((job) => job.jobId !== action.payload);

      return { ...state, jobs: newJobs };
    }
    case ActionTypes.ADD_STATUS_COUNT: {
      return incrementValue(state, action.payload, 1);
    }
    case ActionTypes.UPDATE_STATUS: {
      const [prevStatus, curStatus] = action.payload;

      const state1 = incrementValue(state, prevStatus, -1);
      const state2 = incrementValue(state1, curStatus, 1);

      return state2;
    }
    default:
      throw new Error(`Unknown action ${action.type}`);
  }
}

function incrementValue(state, status, amount) {
  const statusToCountField = {
    [JobStatus.ONGOING]: 'numOngoing',
    [JobStatus.SUCCESS]: 'numSuccessful',
    [JobStatus.ERROR]: 'numFailed',
  };

  if (!statusToCountField[status]) {
    throw new Error(`Unknown job status ${status}`);
  }

  const oldValue = state.statusCounts[statusToCountField[status]];

  return {
    ...state,
    statusCounts: {
      ...state.statusCounts,
      [statusToCountField[status]]: checkAmount(oldValue + amount),
    },
  };
}

function checkAmount(amount) {
  if (amount < 0) {
    throw new Error('Amount is less than 0');
  }

  return amount;
}
