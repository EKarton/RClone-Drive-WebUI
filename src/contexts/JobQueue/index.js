import { createContext, useReducer } from 'react';
import { reducer } from './reducer';

export const InitialState = {
  jobs: [],
  statusCounts: {
    numOngoing: 0,
    numSuccessful: 0,
    numFailed: 0,
  },
};

export const JobQueueContext = createContext(InitialState);

export const JobQueueProvider = ({ children, defaultState = InitialState }) => {
  const [state, dispatch] = useReducer(reducer, defaultState);

  return (
    <JobQueueContext.Provider value={{ state, dispatch }}>
      {children}
    </JobQueueContext.Provider>
  );
};

export * from './actionTypes';
