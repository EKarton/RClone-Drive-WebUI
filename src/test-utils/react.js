import { FileViewerProvider } from 'contexts/FileViewerStore';
import { RCloneInfoStateProvider } from 'contexts/RCloneInfoStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState, initialRCloneInfoState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  const renderedComponent = render(
    <RCloneInfoStateProvider defaultState={initialRCloneInfoState}>
      <FileViewerProvider defaultState={initialFileViewerState}>
        <Router history={history}>{component}</Router>
      </FileViewerProvider>
    </RCloneInfoStateProvider>
  );

  return {
    ...renderedComponent,
    history,
  };
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
