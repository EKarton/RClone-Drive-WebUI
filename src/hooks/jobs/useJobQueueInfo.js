import { useContext } from 'react';
import { JobQueueContext } from '../../contexts/JobQueue/index';

export default function useJobQueueInfo() {
  const { state } = useContext(JobQueueContext);
  return state;
}
