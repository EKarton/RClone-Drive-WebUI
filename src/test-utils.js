import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
