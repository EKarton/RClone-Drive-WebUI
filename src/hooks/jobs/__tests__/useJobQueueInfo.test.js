import { JobQueueProvider } from 'contexts/JobQueue/index';
import { render } from 'test-utils/react';
import useJobQueueInfo from '../useJobQueueInfo';

describe('useJobQueueInfo()', () => {
  it('should return the correct info given it is wrapped in <JobQueueProvider>', () => {
    const MockPage = () => {
      const { jobs, statusCounts } = useJobQueueInfo();

      return (
        <>
          <div>{JSON.stringify(jobs)}</div>
          <div>{JSON.stringify(statusCounts)}</div>
        </>
      );
    };

    const { baseElement } = render(
      <JobQueueProvider>
        <MockPage />
      </JobQueueProvider>
    );

    expect(baseElement).toMatchSnapshot();
  });
});
