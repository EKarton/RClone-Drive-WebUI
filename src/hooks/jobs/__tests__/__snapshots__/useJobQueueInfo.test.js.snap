// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useJobQueueInfo() should return the correct info given it is wrapped in <JobQueueProvider> 1`] = `
<body>
  <div>
    <div>
      []
    </div>
    <div>
      {"numOngoing":0,"numSuccessful":0,"numFailed":0}
    </div>
  </div>
</body>
`;
